nexus/src/db/datastore.rs

```diff
@@ -81,6 +81,24 @@ pub struct DataStore {
     pool: Arc<Pool>,
 }
 
+// Errors which may be returned from the DB transaction within
+// [`DataStore::vpc_update_firewall_rules`].
+#[derive(Debug, thiserror::Error)]
+enum UpdateFirewallRulesError {
+    #[error("Collection not found")]
+    CollectionNotFound,
+    #[error("Pool error: {0}")]
+    Pool(#[from] async_bb8_diesel::PoolError),
+}
+
+// Maps a "diesel error" into a "pool error", which
+// is already contained within the error type.
+impl From<diesel::result::Error> for UpdateFirewallRulesError {
+    fn from(err: diesel::result::Error) -> Self {
+        Self::Pool(PoolError::Connection(ConnectionError::Query(err)))
+    }
+}
+
```

Comment:
Does this mean every query that uses collection insert should implement something like this, or could they share one? My immediate reaction is that the contents look very generic for something with such a specific name.

Reply:
So, this is the PR that introduces this change: https://github.com/oxidecomputer/async-bb8-diesel/pull/9

Inside that PR, this is the trait declaration for txns - the bounds on `transaction` are particularly relevant.
These two `From` impl's satisfy those bounds.

```rust
use diesel::result:Error as DieselError;

#[async_trait]
pub trait AsyncConnection<Conn, ConnErr>: AsyncSimpleConnection<Conn, ConnErr>
where
    Conn: 'static + DieselConnection,
{
    /// Runs the function `f` in an context where blocking is safe.
    ///
    /// Any error may be propagated through `f`, as long as that
    /// error type may be constructed from `ConnErr` (as that error
    /// type may also be generated).
    async fn run<R, E, Func>(&self, f: Func) -> Result<R, E>
    where
        R: Send + 'static,
        E: From<ConnErr> + Send + 'static,
        Func: FnOnce(&mut Conn) -> Result<R, E> + Send + 'static;

    async fn transaction<R, E, Func>(&self, f: Func) -> Result<R, E>
    where
        R: Send + 'static,
        E: From<DieselError> + From<ConnErr> + Send + 'static,
        Func: FnOnce(&mut Conn) -> Result<R, E> + Send + 'static,
    {
        self.run(|conn| conn.transaction(|c| f(c))).await
    }
}
```

I spent a while trying to figure out if I could avoid the `E: From<DieselError>` constraint, and just "shove Diesel errors into connection errors anyway", but I was unsuccessful.

ConnectionError can easily impl `From<DieselError>`
PoolError can easily impl `From<DieselError> + From<ConnectionError>`

but having an automated way to jump from `DieselError` to "your arbitrary error type, which also contains a `PoolError`" is a little... tricky? The problem is that it's basically a two-hop conversion, from `DieselError` -> `PoolError` -> "Your error type", which is `UpdateFirewallRulesError` here.

I could probably reduce this specific example to:

```rust
impl From<diesel::result::Error> for UpdateFirewallRulesError {
    fn from(err: diesel::result::Error) -> Self {
        Self::Pool(PoolError::from(err)))
    }
}
```

which would at least make it a little less verbose?

Reply:
I suppose we could also use generics to resolve this pattern, like:

```rust
enum TxnError<T> {
   CustomError(T),
   PoolError(...)
}
```

to avoid re-defining these conversions?

Reply:
I went ahead and implemented the generics-based version, so hopefully the conversion from the Diesel should only need to be defined in one spot. Lemme know what you think.

Reply:
Great! I think it makes a lot more sense.